window based scrolling

nicholascarroll opened the pull request:
issue #32 independent window scrolling.  It also includes toggle-truncate-lines, POSIX 2001 compliance and use of tilde for home dir in C-x C-f

japanoise commented on find.c:
Any reason we're commenting code out here rather than deleting it?

If we plan to re-add it in future/re-use it for something else, let's get a comment in/maybe a ticket number.

japanoise commented on main.c:
```diff
@@ -1823,9 +1969,9 @@ void editorProcessKeypress(int c) {
-	case CTRL('l'):
-		editorRecenter(bufr);
-		break;
+	case CTRL('l'): {
+		editorRecenter(win);
+	} break;
```
Little curious about this syntax - any reason we can't just:

```suggestion
	case CTRL('l'):
		editorRecenter(win);
		break;
```

japanoise commented on main.c:
```diff
@@ -1867,31 +2013,58 @@ void editorProcessKeypress(int c) {
-		break;
+
+		editorScroll();
+	} break;
```
Same question here about bracing the cases

nicholascarroll replied:
ok i removed them

japanoise commented on main.c:
```diff
@@ -1867,31 +2013,58 @@ void editorProcessKeypress(int c) {
 		bufr->cx = 0;
-		bufr->rowoff = 0;
 		break;
+	case CUSTOM_INFO_MESSAGE: {
+		int winIdx = windowFocusedIdx(&E);
+		struct editorWindow *win = E.windows[winIdx];
+		struct editorBuffer *buf = win->buf;
+
+		editorSetStatusMessage(
+			"(buf->cx%d,cy%d) (win->scx%d,scy%d) win->height=%d screenrows=%d, screencols=%d",
+			buf->cx, buf->cy, win->scx, win->scy, win->height,
+			E.screenrows, E.screencols);
+	} break;
```
Once more for luck :grin: 

```suggestion
	case CUSTOM_INFO_MESSAGE:
		int winIdx = windowFocusedIdx(&E);
		struct editorWindow *win = E.windows[winIdx];
		struct editorBuffer *buf = win->buf;

		editorSetStatusMessage(
			"(buf->cx%d,cy%d) (win->scx%d,scy%d) win->height=%d screenrows=%d, screencols=%d",
			buf->cx, buf->cy, win->scx, win->scy, win->height,
			E.screenrows, E.screencols);
		break;
```

japanoise reviewed (commented):
I'll need to compile and test this but this is looking good. Some minor code quality/nits here.

japanoise reviewed (changes requested):
So there's a small bug here (at least on my terminal - kitty on arch linux). When I split the `txt/stresstest.txt` file into two windows, there's an off-by one in the bottom window. Scrolling as far to the top the cursor goes, it overlaps the status bar of the top window; scrolling to as far to the bottom as the cursor goes, it is one line above the last visible line.

Couldn't get the cursor in the screenshot so enjoy the mspaint :zany_face: 

![2024-07-22-17:57:24](https://github.com/user-attachments/assets/839bf5ca-b131-44dc-a0f0-a7e1439cb29e)

nicholascarroll commented:
Ok let me replicate that and fix it.

nicholascarroll commented:
I fixed the bugs (there was actually 3)  but I changed what constitutes window height and then I realized you may not like that. I made window height not include the status bar height. If you want I can change it back?

nicholascarroll commented:
just found another bug. I'll get back in touch when its ready

japanoise commented:
> I fixed the bugs (there was actually 3) but I changed what constitutes window height and then I realized you may not like that. I made window height not include the status bar just because i thought it was a good idea at the time but now i realize that may be confusing. In Emacs there are two: window-body-height which does not include the mode line and window-total-height which does. If you want I can change it back?

I'm OK with the rendering code being refactored to make hackability better. Change it however is needed to get the functionality working - we can hash out code quality when it's done :smile:

nicholascarroll commented:
OK I think this is finally working properly now.

nicholascarroll commented:
I added an enhancement to switch buffers: it now has the default switch to previous buffer. So C-b RET.

japanoise reviewed (approved):
LGTM
